### PR TITLE
Consensus upgrade

### PR DESCRIPTION
**Pull request by jkitman**
PR for adding consensus upgrades to the API, fixing #948.
- Adds the consensus upgrade signaling code via an authorized API.
- Introduces `ApiEvent` so we can send non-transaction events from the API to consensus thread
- Adds a new integration test for signaling the upgrade
- Eventually the upgrade signaling/start-up should move into an admin UI.

Will follow-up with a refactor suggested by @Maan2003 to introduce `ApiEndpointContext`

### CODE REVIEW

File: client/cli/src/main.rs [hunk 1]
```diff
@@ -331,6 +362,20 @@ enum Command {
     /// Wipe the notes data from the DB. Useful for testing backup & restore
     #[clap(hide = true)]
     WipeNotes,
+
+    /// Signal a consensus upgrade
+    SignalUpgrade {
+        /// Guardian password for authentication
+        password: String,
```
Comment (dpc) [inline client/cli/src/main.rs:369]: Take from the env var via `#[clap(env = "...")]`?

File: client/cli/src/main.rs [hunk 2]
```diff
@@ -331,6 +362,20 @@ enum Command {
...
+        /// Location of the salt file
+        salt_path: PathBuf,
+        /// Peer id of the guardian
+        #[clap(value_parser = parse_peer_id)]
```
Comment (dpc) [inline client/cli/src/main.rs:373]: Probably we should put it into fedimint-core behind a feature flag, gated on optional `clap` dependency. Could be a separate PR.

File: scripts/upgrade-test.sh [hunk 3]
```diff
@@ -0,0 +1,48 @@
...
+server1=$(tail -1 $FM_PID_FILE | head -1)
+
+
+function wait_server_shutdown() {
+  echo "Waiting for $1 to shutdown..."
+  tail --pid=$1 -f /dev/null
+  echo "Server $1 has shutdown."
+}
+
+mine_blocks 110
```
Comment (justinmoon) [inline scripts/upgrade-test.sh:21]: FYI: we mine 101 blocks on bitcoind startup now so this might not be necessary https://github.com/fedimint/fedimint/blob/37a70ba4ef609a456bfb1323ca2a7b598f3b3d1b/fedimint-testing/src/bin/fixtures.rs#L161

File: scripts/upgrade-test.sh [hunk 4]
```diff
@@ -0,0 +1,48 @@
...
+await_fedimint_block_sync
+await_all_peers
+
+# test a consensus upgrade
+$FM_MINT_CLIENT signal-upgrade "pass0" $FM_CFG_DIR/server-0/private.salt 0
+$FM_MINT_CLIENT signal-upgrade "pass1" $FM_CFG_DIR/server-1/private.salt 1
+
+mine_blocks 10
+await_fedimint_block_sync
+await_all_peers
```
Comment (justinmoon) [inline scripts/upgrade-test.sh:31]: Why do we do this instead of just having `server-2` signal right away? Comment would be nice ...

File: client/cli/src/main.rs (hunk 2 above)
Comment (jkitman) [reply client/cli/src/main.rs:373]: Yep, I just put the function next to the other ones.

File: flake.crane.nix [hunk 5]
```diff
@@ -269,6 +269,13 @@ rec {
     buildPhaseCargoCommand = "patchShebangs ./scripts ; ./scripts/reconnect-test.sh";
   });
 
+  cliTestUpgrade = craneLib.mkCargoDerivation (commonCliTestArgs // {
```
Comment (dpc) [inline flake.crane.nix:272]: You need to add this to one more place in this file if you want it run. And then to `./script/test-ci-all.sh` to trigger from the CI.

File: fedimint-core/src/api.rs [hunk 6]
```diff
@@ -775,6 +775,30 @@ impl<C> WsFederationApi<C> {
...
+
+    pub async fn signal_upgrade(&self) -> FederationResult<()> {
+        self.inner
+            .request_current_consensus(
+                "/upgrade".to_owned(),
+                ApiRequestErased::default().with_auth(&self.auth),
+            )
+            .await
+    }
+}
```
Comment (justinmoon) [inline fedimint-core/src/api.rs:800]: nit: would be easier to review if these auth changes were in a separate commit from the signalling changes

File: scripts/upgrade-test.sh (hunk 4 above)
Comment (jkitman) [reply scripts/upgrade-test.sh:31]: True, I don't need it there.

File: fedimint-server/src/consensus/mod.rs [hunk 7]
```diff
@@ -81,6 +82,13 @@ pub struct ConsensusProposal {
     pub force_new_epoch: bool,
 }
 
+/// Events that can be sent from the API to consensus thread
+#[derive(Debug, Clone, Eq, Hash, PartialEq)]
+pub enum ApiEvent {
+    Transaction(Transaction),
+    UpgradeSignal,
+}
+
```
Comment (justinmoon) [inline fedimint-server/src/consensus/mod.rs:91]: nit: would be easier to review if this refactor was separate commit

File: flake.crane.nix (hunk 5 above)
Comment (jkitman) [reply flake.crane.nix:272]: Where was the other place in the file?  I didn't see that for the other tests

File: fedimint-server/src/net/api.rs [hunk 8]
```diff
@@ -266,5 +266,16 @@ fn server_endpoints() -> Vec<ApiEndpoint<FedimintConsensus>> {
                 Ok(fedimint.get_config_with_sig(dbtx).await)
             }
         },
+        api_endpoint! {
+            "/upgrade",
+            async |fedimint: &FedimintConsensus, _dbtx, _v: (), has_auth| -> () {
+                if has_auth {
+                    fedimint.signal_upgrade().await.map_err(|_| ApiError::server_error("Unable to send signal to server".to_string()))?;
```
Comment (justinmoon) [inline fedimint-server/src/net/api.rs:273]: If we're able to send over this channel, we know that the signaling succeeded? We don't need to wait for any response from `fedimintd`?

Comment (jkitman) [reply fedimint-server/src/net/api.rs:273]: Yes it means the `ApiEvent` is in channel that triggers an epoch with the signal in the consensus thread.

Comment (jkitman) [reply fedimint-server/src/net/api.rs:273]: We use the same channel for users who submit txs.

File: flake.crane.nix (hunk 5 above)
Comment (dpc) [reply flake.crane.nix:272]:
```
          cli-test = {
            reconnect = craneBuildNative.cliTestReconnect;
            latency = craneBuildNative.cliTestLatency;
            cli = craneBuildNative.cliTestCli;
            rust-tests = craneBuildNative.cliRustTests;
            always-fail = craneBuildNative.cliTestAlwaysFail;
          };
```

File: fedimint-server/src/net/api.rs (hunk 8 above)
Comment (justinmoon) [reply fedimint-server/src/net/api.rs:273]: Cool. Thanks for explaining.

File: fedimint-core/src/module/mod.rs [hunk 9]
```diff
@@ -172,8 +188,9 @@ macro_rules! __api_endpoint {
                 $state: &'a Self::State,
                 $dbtx: &'a mut fedimint_core::db::ModuleDatabaseTransaction<'b, ModuleInstanceId>,
                 $param: Self::Param,
-                has_auth: bool,
+                _has_auth: bool,
             ) -> ::std::result::Result<Self::Response, $crate::module::ApiError> {
+                $(let $auth = _has_auth;)?
```
Comment (elsirion) [inline fedimint-core/src/module/mod.rs:193]: This looks like a hack. Why not just bind the bool arg to `$auth` if present and `_` otherwise?

**Review by elsirion — APPROVED** [review]

File: fedimint-core/src/module/mod.rs (hunk 9 above)
Comment (jkitman) [reply fedimint-core/src/module/mod.rs:193]: Will get rid of this in a follow-up PR, since we want to put `auth` and `dbtx` into a new `ApiEndpointContext` struct.

File: scripts/upgrade-test.sh [hunk 10]
```diff
@@ -0,0 +1,45 @@
...
+  echo "Waiting for $1 to shutdown..."
+  tail --pid=$1 -f /dev/null
+  echo "Server $1 has shutdown."
+}
+
+await_fedimint_block_sync
+
+# test a consensus upgrade
+FM_PASSWORD="pass0" $FM_MINT_CLIENT signal-upgrade $FM_CFG_DIR/server-0/private.salt 0
+FM_PASSWORD="pass1" $FM_MINT_CLIENT signal-upgrade $FM_CFG_DIR/server-1/private.salt 1
```
[the remark this answers is not in the document]
Comment (jkitman) [reply scripts/upgrade-test.sh:25]: I think the best security option is to put the salt in the admin UI.  Failing that, it would live on fedimintd and not be passed.

[the remark this answers is not in the document]
Comment (dpc) [reply scripts/upgrade-test.sh:25]: Make the `salt == hash(federationid || peer_id)` or something? AFAIK it doesn't have to be secret, just different to prevent rainbow tables on some other forms of mass-breaking of keys, no?

File: fedimint-server/src/lib.rs [hunk 11]
```diff
@@ -496,7 +496,7 @@ impl FedimintServer {
         }
 
         tokio::select! {
-            _peek = Pin::new(&mut self.tx_receiver).peek() => Ok(EpochTriggerEvent::NewTransaction),
+            _peek = Pin::new(&mut self.api_receiver).peek() => Ok(EpochTriggerEvent::NewTransaction),
```
Comment (Maan2003) [inline fedimint-server/src/lib.rs:499]: no longer correct?

Comment (jkitman) [reply fedimint-server/src/lib.rs:499]: How so?  This is just a rename since we can receive any `ApiEvent`

File: scripts/upgrade-test.sh (hunk 10 above)
[the remark this answers is not in the document]
Comment (jkitman) [reply scripts/upgrade-test.sh:25]: It doesn't have to be secret, although if you make it secret it does help users with low-entropy passwords.

Although in all likelihood it's probably more of a foot-gun to depend on users to back it up properly so maybe we should just derive it.

File: fedimint-server/src/lib.rs (hunk 11 above)
Comment (Maan2003) [reply fedimint-server/src/lib.rs:499]: is `EpochTriggerEvent::NewTransaction` correct here?

Comment (jkitman) [reply fedimint-server/src/lib.rs:499]: Ah I see, yes that should be renamed too.

**Review by Maan2003 — APPROVED** [review]
